File: reference/internal/ui/components/statepicker.go
package components

import (
	"fmt"

	"github.com/Elpulgo/azdo/internal/azdevops"
	"github.com/Elpulgo/azdo/internal/ui/styles"
	"github.com/charmbracelet/bubbles/key"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

// StateSelectedMsg is sent when a state option is selected
type StateSelectedMsg struct {
	State string
}

// stateOption represents a single state choice in the picker
type stateOption struct {
	Name      string
	Icon      string
	IsCurrent bool
}

// StatePicker is a modal component for selecting a work item state
type StatePicker struct {
	styles       *styles.Styles
	visible      bool
	width        int
	height       int
	options      []stateOption
	cursor       int
	currentState string
}

// NewStatePicker creates a new state picker
func NewStatePicker(s *styles.Styles) StatePicker {
	return StatePicker{
		styles:  s,
		visible: false,
		cursor:  0,
	}
}

// SetStates sets the available states and positions the cursor on the current state
func (sp *StatePicker) SetStates(states []azdevops.WorkItemTypeState, currentState string) {
	sp.currentState = currentState
	sp.options = make([]stateOption, len(states))
	sp.cursor = 0

	for i, s := range states {
		sp.options[i] = stateOption{
			Name:      s.Name,
			Icon:      stateIcon(s.Category),
			IsCurrent: s.Name == currentState,
		}
		if s.Name == currentState {
			sp.cursor = i
		}
	}
}

// Show makes the state picker visible
func (sp *StatePicker) Show() {
	sp.visible = true
}

// Hide makes the state picker invisible
func (sp *StatePicker) Hide() {
	sp.visible = false
}

// IsVisible returns whether the state picker is visible
func (sp StatePicker) IsVisible() bool {
	return sp.visible
}

// SetSize sets the dimensions for centering
func (sp *StatePicker) SetSize(width, height int) {
	sp.width = width
	sp.height = height
}

// GetCursor returns the current cursor position
func (sp StatePicker) GetCursor() int {
	return sp.cursor
}

// Update handles messages
func (sp StatePicker) Update(msg tea.Msg) (StatePicker, tea.Cmd) {
	if !sp.visible {
		return sp, nil
	}

	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch {
		case key.Matches(msg, key.NewBinding(key.WithKeys("esc", "q"))):
			sp.visible = false
			return sp, nil

		case key.Matches(msg, key.NewBinding(key.WithKeys("up", "k"))):
			if sp.cursor > 0 {
				sp.cursor--
			}
			return sp, nil

		case key.Matches(msg, key.NewBinding(key.WithKeys("down", "j"))):
			if sp.cursor < len(sp.options)-1 {
				sp.cursor++
			}
			return sp, nil

		case key.Matches(msg, key.NewBinding(key.WithKeys("enter"))):
			if len(sp.options) == 0 {
				return sp, nil
			}
			selected := sp.options[sp.cursor]
			sp.visible = false
			return sp, func() tea.Msg {
				return StateSelectedMsg{State: selected.Name}
			}
		}
	}

	return sp, nil
}

// View renders the state picker
func (sp StatePicker) View() string {
	if !sp.visible {
		return ""
	}

	titleText := "Change Work Item State"
	helpTextStr := "↑/↓: navigate • enter: select • esc/q: cancel"

	maxWidth := minModalWidth
	if len(titleText) > maxWidth {
		maxWidth = len(titleText)
	}
	if len(helpTextStr) > maxWidth {
		maxWidth = len(helpTextStr)
	}

	for _, opt := range sp.options {
		label := opt.Name
		if opt.IsCurrent {
			label += " (current)"
		}
		lineLen := len(fmt.Sprintf("> %s %s", opt.Icon, label))
		if lineLen > maxWidth {
			maxWidth = lineLen
		}
	}

	var optionList string
	for i, opt := range sp.options {
		cursor := " "
		if i == sp.cursor {
			cursor = ">"
		}

		label := opt.Name
		if opt.IsCurrent {
			label += " (current)"
		}

		line := fmt.Sprintf("%s %s %s", cursor, opt.Icon, label)

		if i == sp.cursor {
			line = lipgloss.NewStyle().
				Foreground(sp.styles.Theme.GetSelectForeground()).
				Background(sp.styles.Theme.GetSelectBackground()).
				Width(maxWidth).
				Render(line)
		} else {
			line = lipgloss.NewStyle().
				Foreground(sp.styles.Theme.GetForeground()).
				Background(sp.styles.Theme.GetBackground()).
				Width(maxWidth).
				Render(line)
		}

		optionList += line + "\n"
	}

	title := lipgloss.NewStyle().
		Foreground(sp.styles.Theme.GetPrimary()).
		Background(sp.styles.Theme.GetBackground()).
		Bold(true).
		Width(maxWidth).
		Render(titleText)

	helpText := lipgloss.NewStyle().
		Foreground(sp.styles.Theme.GetForegroundMuted()).
		Background(sp.styles.Theme.GetBackground()).
		Width(maxWidth).
		Render(helpTextStr)

	content := lipgloss.JoinVertical(
		lipgloss.Left,
		title,
		"",
		optionList,
		helpText,
	)

	modalStyle := lipgloss.NewStyle().
		Border(lipgloss.RoundedBorder()).
		BorderForeground(sp.styles.Theme.GetBorder()).
		Padding(1, 2).
		Background(sp.styles.Theme.GetBackground())

	modal := modalStyle.Render(content)

	if sp.width > 0 && sp.height > 0 {
		modal = lipgloss.Place(
			sp.width,
			sp.height,
			lipgloss.Center,
			lipgloss.Center,
			modal,
		)
	}

	return modal
}

// stateIcon returns an icon for the work item state category
func stateIcon(category string) string {
	switch category {
	case "Proposed":
		return "○"
	case "InProgress":
		return "◐"
	case "Resolved":
		return "●"
	case "Completed":
		return "✓"
	default:
		return "○"
	}
}

File: src/Azdo.Tui/Components/StatePicker.cs
using Azdo.Tui.Rendering;
using Azdo.Tui.Runtime;
using StyleSet = Azdo.Tui.Styles.Styles;

namespace Azdo.Tui.Components;

/// <summary>Emitted when a state option is selected (≈ <c>components.StateSelectedMsg</c>).</summary>
public sealed record StateSelectedMsg(string State) : IMsg;

/// <summary>
/// An available work item state passed to <see cref="StatePicker.SetStates"/>. Kept
/// local to avoid a dependency on Core's <c>WorkItemTypeState</c>.
/// </summary>
public readonly record struct WorkItemStateOption(string Name, string Category);

/// <summary>A single state choice in the picker.</summary>
internal sealed class StateOption
{
    public string Name = "";
    public string Icon = "";
    public bool IsCurrent;
}

/// <summary>A modal for selecting a work item state (≈ <c>components.StatePicker</c>).</summary>
public sealed class StatePicker
{
    private const int MinModalWidth = 80;

    private readonly StyleSet _styles;
    private bool _visible;
    private int _width;
    private int _height;
    private List<StateOption> _options = new();
    private int _cursor;
    private string _currentState = "";

    public StatePicker(StyleSet styles) => _styles = styles;

    /// <summary>Sets the available states and positions the cursor on the current state.</summary>
    public void SetStates(IEnumerable<WorkItemStateOption> states, string currentState)
    {
        _currentState = currentState;
        _cursor = 0;
        var src = states.ToList();
        _options = new List<StateOption>(src.Count);
        for (int i = 0; i < src.Count; i++)
        {
            var s = src[i];
            _options.Add(new StateOption
            {
                Name = s.Name,
                Icon = StateIcon(s.Category),
                IsCurrent = s.Name == currentState,
            });
            if (s.Name == currentState) _cursor = i;
        }
    }

    public void Show() => _visible = true;
    public void Hide() => _visible = false;
    public bool IsVisible => _visible;
    public void SetSize(int width, int height) { _width = width; _height = height; }
    public int GetCursor() => _cursor;

    public Cmd? Update(IMsg msg)
    {
        if (!_visible) return null;
        if (msg is KeyMsg key)
        {
            switch (key.Key)
            {
                case "esc":
                case "q":
                    _visible = false;
                    return null;
                case "up":
                case "k":
                    if (_cursor > 0) _cursor--;
                    return null;
                case "down":
                case "j":
                    if (_cursor < _options.Count - 1) _cursor++;
                    return null;
                case "enter":
                    if (_options.Count == 0) return null;
                    var selected = _options[_cursor];
                    _visible = false;
                    return Commands.Of(new StateSelectedMsg(selected.Name));
            }
        }
        return null;
    }

    public string View()
    {
        if (!_visible) return "";

        var theme = _styles.Theme;
        const string titleText = "Change Work Item State";
        const string helpTextStr = "↑/↓: navigate • enter: select • esc/q: cancel";

        int maxWidth = MinModalWidth;
        if (titleText.Length > maxWidth) maxWidth = titleText.Length;
        if (helpTextStr.Length > maxWidth) maxWidth = helpTextStr.Length;

        foreach (var opt in _options)
        {
            string label = opt.IsCurrent ? opt.Name + " (current)" : opt.Name;
            int lineLen = $"> {opt.Icon} {label}".Length;
            if (lineLen > maxWidth) maxWidth = lineLen;
        }

        var optionList = new System.Text.StringBuilder();
        for (int i = 0; i < _options.Count; i++)
        {
            var opt = _options[i];
            string cursor = i == _cursor ? ">" : " ";
            string label = opt.IsCurrent ? opt.Name + " (current)" : opt.Name;
            string line = $"{cursor} {opt.Icon} {label}";

            line = i == _cursor
                ? Style.New().Foreground(theme.SelectForeground).Background(theme.SelectBackground).Width(maxWidth).Render(line)
                : Style.New().Foreground(theme.Foreground).Background(theme.Background).Width(maxWidth).Render(line);

            optionList.Append(line).Append('\n');
        }

        string title = Style.New()
            .Foreground(theme.Primary)
            .Background(theme.Background)
            .Bold()
            .Width(maxWidth)
            .Render(titleText);

        string helpText = Style.New()
            .Foreground(theme.ForegroundMuted)
            .Background(theme.Background)
            .Width(maxWidth)
            .Render(helpTextStr);

        string content = Layout.JoinVertical(HAlign.Left, title, "", optionList.ToString(), helpText);

        var modalStyle = Style.New()
            .WithBorder(Border.Rounded)
            .BorderForeground(theme.Border)
            .Padding(1, 2)
            .Background(theme.Background);

        var modal = modalStyle.Render(content);

        if (_width > 0 && _height > 0)
            modal = Layout.Place(_width, _height, HAlign.Center, VAlign.Center, modal);

        return modal;
    }

    /// <summary>Returns an icon for the work item state category.</summary>
    private static string StateIcon(string category) => category switch
    {
        "Proposed" => "○",
        "InProgress" => "◐",
        "Resolved" => "●",
        "Completed" => "✓",
        _ => "○",
    };
}
